dashboard/src/actions/charts.ts

```diff
@@ -47,35 +84,110 @@ export const errorReadme = createAction("ERROR_README", resolve => {
 });
 
 const allActions = [
-  requestCharts,
-  requestChart,
   errorChart,
-  receiveCharts,
+  errorReadme,
+  reachEnd,
+  receiveChartCategories,
   receiveChartVersions,
-  selectChartVersion,
-  requestDeployedChartVersion,
+  receiveCharts,
+  receiveChartsSearch,
   receiveDeployedChartVersion,
+  requestChart,
+  requestCharts,
+  requestChartsCategories,
+  requestChartsSearch,
+  requestChartsVersions,
+  requestDeployedChartVersion,
   resetChartVersion,
+  resetChartsSearch,
+  resetPaginaton,
+  selectChartVersion,
   selectReadme,
-  errorReadme,
 ];
 
 export type ChartsAction = ActionType<typeof allActions[number]>;
 
-export function fetchCharts(
+export function fetchChartsWithPagination(
   cluster: string,
   namespace: string,
-  repo: string,
-): ThunkAction<Promise<void>, IStoreState, null, ChartsAction> {
+  repos: string,
+  query: string,
+  page: number,
+  size: number,
+  nextPage: number,
+): ThunkAction<Promise<IChart[]>, IStoreState, null, ChartsAction> {
   return async dispatch => {
-    dispatch(requestCharts());
     try {
-      const charts = await Chart.fetchCharts(cluster, namespace, repo);
-      if (charts) {
-        dispatch(receiveCharts(charts));
+      if (page === nextPage && !(query === "" && size === 0)) {
+        dispatch(requestCharts(page)); // set the current processing page
+        const charts = await Chart.fetchChartsWithPagination(
+          cluster,
+          namespace,
+          repos,
+          query,
+          page,
+          size,
+        );
+        if (charts && charts.length > 0) {
+          dispatch(receiveCharts(charts));
+        } else {
+          dispatch(reachEnd());
+        }
+        return charts;
+      } else {
+        return [];
       }
     } catch (e) {
       dispatch(errorChart(new FetchError(e.message)));
+      return [];
+    }
+  };
+}
+
+export function fetchChartsSearch(
+  cluster: string,
+  namespace: string,
+  repos: string,
+  query: string,
+  page: number,
+  size: number,
+): ThunkAction<Promise<IChart[]>, IStoreState, null, ChartsAction> {
+  return async dispatch => {
+    dispatch(requestChartsSearch(query));
+    try {
+      const charts = await Chart.fetchChartsWithPagination(
+        cluster,
+        namespace,
+        repos,
+        query,
+        page,
+        size,
+      );
+      dispatch(receiveChartsSearch(charts, query));
+      return charts;
+    } catch (e) {
+      dispatch(errorChart(new FetchError(e.message)));
+      return [];
+    }
+  };
+}
+
+export function fetchChartCategories(
+  cluster: string,
+  namespace: string,
+  repos: string,
+): ThunkAction<Promise<IChartCategory[]>, IStoreState, null, ChartsAction> {
+  return async dispatch => {
+    dispatch(requestChartsCategories());
+    try {
+      const categories = await Chart.fetchChartCategories(cluster, namespace, repos);
+      if (categories) {
+        dispatch(receiveChartCategories(categories));
```

Comment:
I always find it odd that we don't dispatch a receive when we receive zero categories (ie `[]`), but I think it's what other actions do.

Reply:
Yes, I just tried to replicate the behavior of other actions in this case